openwpm/storage/arrow_storage.py

```diff
@@ -97,63 +96,76 @@ def _is_cache_full(self) -> bool:
         """
         if interrupted:
             await self.store_record(INCOMPLETE_VISITS, visit_id, {"visit_id": visit_id})
-        assert self.storing_lock is not None
-        assert self.flush_event is not None
         # This code is pretty tricky as there are a number of things going on
-        # 1. No finalize_visit_id shouldn't return unless the visit has been saved to storage
+        # 1. The awaitable returned by finalize_visit_id should only
+        #    resolve once the data is saved to persistent storage
         # 2. No new batches should be created while saving out all the batches
         async with self.storing_lock:
-            # After flush_cache has executed the event needs to be rearmed
-            # so that newly created wait_on_condition don't just complete
-            # instantly
-            self.flush_event.clear()
             self._create_batch(visit_id)
+
+            event = asyncio.Event()
+            self.flush_events.append(event)
+
             if self._is_cache_full():
                 await self.flush_cache(self.storing_lock)
 
-            async def wait_on_condition(event: asyncio.Event) -> None:
-                await event.wait()
+            async def wait_on_condition(e: asyncio.Event) -> None:
+                await e.wait()
 
-            return wait_on_condition(self.flush_event)
+            return asyncio.create_task(wait_on_condition(event))
 
     @abstractmethod
     async def write_table(self, table_name: TableName, table: Table) -> None:
         """Write out the table to persistent storage
         This should only return once it's actually saved out
         """
 
-    async def flush_cache(self, cond: asyncio.Lock = None) -> None:
+    async def flush_cache(self, lock: asyncio.Lock = None) -> None:
         """We need to hack around the fact that asyncio has no reentrant lock
-        So we either grab the storing storing_lock ourselves or the caller needs
+        So we either grab the storing_lock ourselves or the caller needs
         to pass us the locked storing_lock
         """
-        assert self.storing_lock is not None
-        assert self.flush_event is not None
-        got_cond = cond is not None
-        if not got_cond:
-            cond = self.storing_lock
-            await cond.acquire()
-        assert cond is not None and cond.locked()
+        got_lock = lock is not None
+        if not got_lock:
+            lock = self.storing_lock
+            await lock.acquire()
+
+        assert lock == self.storing_lock and lock.locked()
+
         for table_name, batches in self._batches.items():
             table = pa.Table.from_batches(batches)
             await self.write_table(table_name, table)
-        self.flush_event.set()
-        if not got_cond:
-            cond.release()
+        self._batches.clear()
+
+        for event in self.flush_events:
+            event.set()
+        self.flush_events.clear()
+
+        if not got_lock:
```

Comment:
```suggestion
        if not has_lock_arg:
```

I've updated the arg name here because I was a little confused by `got_lock` (i.e. after we locked locally I assumed we'd expect this to be true).